### PR TITLE
langchain-mongodb: Add MongoDB LLM Cache

### PR DESCRIPTION
**Pull request by Jibola**
# Description

- **Description:** Adding MongoDB LLM Caching Layer abstraction
- **Issue:** N/A
- **Dependencies:** None
- **Twitter handle:** @mongodb

Checklist:

- [x] PR title: Please title your PR "package: description", where "package" is whichever of langchain, community, core, experimental, etc. is being modified. Use "docs: ..." for purely docs changes, "templates: ..." for template changes, "infra: ..." for CI changes.
  - Example: "community: add foobar LLM"
- [x] PR Message (above)
- [x] Pass lint and test: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified to check that you're passing lint and testing. See contribution guidelines for more information on how to write/run tests, lint, etc: https://python.langchain.com/docs/contributing/
- [ ] Add tests and docs: If you're adding a new integration, please include
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of @baskaryan, @efriis, @eyurtsev, @hwchase17.

### CODE REVIEW

File: libs/community/langchain_community/cache.py [hunk 1]
```diff
@@ -19,12 +19,14 @@
 
     BaseCache --> <name>Cache  # Examples: InMemoryCache, RedisCache, GPTCache
 """
+
```
Comment (Jibola) [inline libs/community/langchain_community/cache.py:22]:
```suggestion
```

File: libs/community/langchain_community/vectorstores/mongodb_atlas.py [hunk 2]
```diff
@@ -204,6 +204,8 @@ def _similarity_search_with_score(
         ]
         if post_filter_pipeline is not None:
             pipeline.extend(post_filter_pipeline)
+        from pprint import pprint
+        pprint({k:v for k,v in params.items() if k != "queryVector"})
```
Comment (Jibola) [inline libs/community/langchain_community/vectorstores/mongodb_atlas.py:208]:
```suggestion
```

**Comment by Jibola** [comment]
cc: @ShaneHarvey @blink1073

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

**Comment by Jibola** [comment]
cc: @prakul

File: libs/community/langchain_community/cache.py [hunk 3]
```diff
@@ -1836,10 +1838,215 @@ async def adelete_by_document_id(self, document_id: str) -> None:
...
+        if search_response:
+            return_val = search_response[0][0].metadata.get("return_val")
+            return _loads_generations(return_val) or return_val
+
+    def update(
+        self,
+        prompt: str,
+        llm_string: str,
+        return_val: RETURN_VAL_TYPE,
+        wait_until_ready: Optional[bool] = None,
```
[the remark this answers is not in the document]
Comment (Jibola) [reply libs/community/langchain_community/cache.py:2023]: how about `ensure_indexed`?
or `wait_until_indexed`?

[the remark this answers is not in the document]
Comment (blink1073) [reply libs/community/langchain_community/cache.py:2023]: I like `wait_until_indexed`

**Review by efriis — COMMENTED** [review]
Jib will move to partner package.

Some comments
- can we add some unit tests
- can we add docs to the caching page https://python.langchain.com/docs/modules/model_io/chat/chat_model_caching

**Comment by Jibola** [comment]
> Jib will move to partner package.
> 
> Some comments
> 
> * can we add some unit tests
> * can we add docs to the caching page https://python.langchain.com/docs/modules/model_io/chat/chat_model_caching

Yeah, I've decided I'll move these changes to the partners/mongodb-atlas once the [partner-package PR is in ](https://github.com/langchain-ai/langchain/pull/17652).

**Review by efriis — APPROVED** [review]

**Pull request merged**